### PR TITLE
Closes #794 display the smaller size on image library data

### LINKED ISSUE #794: In media library New image size shall be that of AVIF|webP|optimized based on which exists
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version => Avif commit: https://github.com/wp-media/imagify-plugin/commit/b396de27cda71d9b16e311a0f55f482bebf13380
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
The New image size is always the original image optimized size instead of the webP (as on trunk) or the AVIF. After discussion, with @piotrbak , agreed that we shall display the new size as the smallest size

**To Reproduce**
1. Enable the Avif
2. upload image
3. check the image in the media library => compare the new size to that in ftp, it will be the optimized image size, not the AVIF one although AVIF is smaller

**Expected behavior**
The smaller size of the optimized image shall be displayed in the media library as a new size

**Screenshots**
Screenshot from 2024-02-02 12-50-19

**Additional context**
- on trunk, if create webP is enabled, then we will display the webP size which is < original optimized image size (so sounds like regression on PR)

**Acceptance Criteria (for WP Media team use only)**
Clear instructions for developers, to be added before the grooming
- New size is always = smallest size version of image , either original optimized, .webP, or .AVIF based on which version is available

### PR DESCRIPTION
**Pull request by Khadreal**
# Description
Display the smaller size of the image(avif)

Fixes #794 

### User documentation

This will show the avif image size & percentage when available 

### Technical documentation

*Explain how this code works. Diagram & drawings are welcomed.*

## Type of change
- [x] Enhancement (non-breaking change which improves an existing functionality).

## New dependencies
N/A

## Risks
*List possible performance & security issues or risks, and explain how they have been mitigated.*

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.
- [x] I did not introduce unecessary complexity.

### CODE REVIEW

**Review by Miraeld — APPROVED** [review]

**Review by remyperona — CHANGES_REQUESTED** [review]
I think we should go further on this one, updating this method and other methods in the class that reference webp, to make it work with the new implementation (using next-gen). This would also require to update the interface.

**Comment by Khadreal** [comment]
> I think we should go further on this one, updating this method and other methods in the class that reference webp, to make it work with the new implementation (using next-gen). This would also require to update the interface.

If I understand correctly, we should update the method from webp to nextgen ?

**Comment by remyperona** [comment]
Yes, update all occurrences of webp to nextgen. I see there is also a use of `imagify_path_to_webp()` that should be updated in this class for example.

**Review by Miraeld — APPROVED** [review]

File: classes/Optimization/Data/AbstractData.php [hunk 1]
```diff
@@ -281,32 +281,38 @@ public function get_original_size( $human_format = true, $decimals = 2 ) {
...
 	 */
-	public function get_optimized_size( $human_format = true, $decimals = 2, $use_webp = true ) {
+	public function get_optimized_size( $human_format = true, $decimals = 2, $use_nextgen = true ) {
 		if ( ! $this->is_valid() ) {
 			return $human_format ? imagify_size_format( 0, $decimals ) : 0;
 		}
 
 		$data  = $this->get_optimization_data();
 		$media = $this->get_media();
+		$format = '.webp';
```
Comment (remyperona) [inline classes/Optimization/Data/AbstractData.php]:
```suggestion
		$format = 'webp';
```

File: classes/Optimization/Data/AbstractData.php [hunk 2]
```diff
@@ -281,32 +281,38 @@ public function get_original_size( $human_format = true, $decimals = 2 ) {
...
 
-		if ( $use_webp ) {
+		if ( $use_nextgen ) {
 			$process_class_name = imagify_get_optimization_process_class_name( $media->get_context() );
-			$webp_size_name     = 'full' . constant( $process_class_name . '::WEBP_SUFFIX' );
+			$nextgen_size_name     = 'full' . constant( $process_class_name . '::WEBP_SUFFIX' );
+
+			if ( get_imagify_option( 'convert_to_avif' ) ) {
+				$nextgen_size_name     = 'full' . constant( $process_class_name . '::AVIF_SUFFIX' );
+				$format = '.avif';
```
Comment (remyperona) [inline classes/Optimization/Data/AbstractData.php]:
```suggestion
				$format = 'avif';
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Comment by Mai-Saad** [comment]
@Khadreal Thanks for the PR. please find exploratory note below:
- when upload image while avif is off, we arenot generating webP by default
- when enable avif after that, we arenot generating avif for uploaded images in previous step
can you please check? (those points were working on the main feature branch)

**Comment by Khadreal** [comment]
> @Khadreal Thanks for the PR. please find exploratory note below:
> 
> * when upload image while avif is off, we arenot generating webP by default
> * when enable avif after that, we arenot generating avif for uploaded images in previous step
>   can you please check? (those points were working on the main feature branch)

The branch has been updated to the latest feature branch

**Review by jeawhanlee — APPROVED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@Khadreal Thanks for the PR. working as in trunk after merging main branch to the PR locally while image have only webp.

Note: we have some discussion about sizes [here](https://wp-media.slack.com/archives/CU0F6EGQ1/p1709458320636499) (probably it isnot always the smallest on trunk), however, the agreed expected shall be in separate GH as it's the case on trunk.

**Update:** as per slack discussion, we shall display here the size of avif , if not exist then webp, if not exist then optimized => now we display in media library new size as per enabled option. which doesn't meet this update from slack. @Khadreal can you please check?
steps:
1- avif enabled
2- upload image => avif created and new size =that of avif
3- disable avif => webp created
4- check media library => new size = webp size instead of avif
bird-1905255_1280

/cc @piotrbak

**Review by remyperona — APPROVED** [review]

File: classes/Bulk/Bulk.php [hunk 3]
```diff
@@ -597,20 +598,18 @@ public function bulk_get_stats_callback() {
...
+			return;
+		}
+
+		if ( ! isset( $old_value['convert_to_avif'] ) && ! isset( $value['convert_to_avif'] ) ) {
 			return;
 		}
 
-		if ( empty( $value['convert_to_avif'] ) ) {
-			// new value is disabled, do nothing.
+		if ( isset( $value['convert_to_avif'] ) && ! (bool) $value['convert_to_avif'] ) {
```
Comment (remyperona) [inline classes/Bulk/Bulk.php]: the old code is the correct one, not this one

**Review by remyperona — CHANGES_REQUESTED** [review]

**Comment by Khadreal** [comment]
This PR https://github.com/wp-media/imagify-plugin/pull/851 once merge will fixed the failing test error

**Comment by MathieuLamiot** [comment]
@Khadreal should this go back to ready for review, or there are still work needed here?

**Comment by Khadreal** [comment]
> @Khadreal should this go back to ready for review, or there are still work needed here?

I'll move it, done with the work but didn't move it because of the test failing

**Comment by Mai-Saad** [comment]
@Khadreal Thanks for the update. Please find exploratory note below:
- Can see the following error for medium size image 
`[14-Mar-2024 12:29:32 UTC] PHP Warning:  Undefined array key "optimized_size" in /home/mai/Local Sites/imagifynginx6/app/public/wp-content/plugins/imagify-plugin/classes/Optimization/Data/AbstractData.php on line 309`
in the following scenario:
1- install and activate imagify
2- upload med size image attached
marchmed1
3- enable avif 

or 
1- activate avif 
2- upload the same image

Screenshot from 2024-03-14 14-24-53


can you please check?

Note: in general, related test case https://wpmediaqa.testrail.io/index.php?/cases/view/15271

**Comment by Khadreal** [comment]
Got this error, wondering if you're seeing something like this as well. This is when avif is not enable.

Screenshot 2024-03-15 at 15 33 51

**Comment by Khadreal** [comment]
> #794



> Got this error, wondering if you're seeing something like this as well. This is when avif is not enable.
> 
> Screenshot 2024-03-15 at 15 33 51

My last update fixed this too

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@Khadreal Thanks for the update, fixed in both scenarios mentioned above but can see problem if we refresh admin after scenario 2 i.e
1- install and activate imagify
2- enable avif
3- upload the image => avif created and no error
4- refresh media page , the warning is there

in log can see
`[18-Mar-2024 09:52:12 UTC] PHP Warning:  Undefined array key "full@imagify-webp" in /home/mai/Local Sites/imagifynginx6/app/public/wp-content/plugins/imagify-plugin/classes/Optimization/Data/AbstractData.php on line 314`

note: it happens with any optimized image not just the previous one
[screen-capture (88).webm](https://github.com/wp-media/imagify-plugin/assets/76941962/9fb5f647-d695-4d0c-ac29-6a2232d79ed5)

**Comment by Khadreal** [comment]
Fixed. please check again. 
Sorry for the back and forth

**Comment by Mai-Saad** [comment]
@Khadreal Thanks for the update. Working fine now with the following note (which is fine as per https://wp-media.slack.com/archives/CU0F6EGQ1/p1709546156479809?thread_ts=1709458320.636499&cid=CU0F6EGQ1)
- if we manually deleted next-gen, the displayed size won't change even after refresh media page (i.e displayed size is that of avif, then manually delete the avif, the displayed size still is that of avif)

**Pull request merged**